src/flb_output.c

```diff
@@ -173,6 +173,98 @@ void flb_output_coro_add(struct flb_output_instance *ins, struct flb_coro *coro)
     mk_list_add(&out_flush->_head, &ins->flush_list);
 }
 
+/*
+ * Queue a task to be flushed at a later time
+ */
+static int flb_output_task_queue_enqueue(struct flb_task_queue *queue,
+                                         struct flb_task *task,
+                                         struct flb_output_instance *out_ins,
+                                         struct flb_config *config)
+{
+    struct flb_task_enqueued *queued_task;
+
+    queued_task = flb_malloc(sizeof(struct flb_task_enqueued));
+    if (!queued_task) {
+        flb_errno();
+        return -1;
+    }
+    queued_task->out_instance = out_ins;
+    queued_task->task = task;
+    queued_task->config = config;
+
+    mk_list_add(&queued_task->_head, &queue->pending);
+    return 0;
+}
+
+/*
+ * Pop task from pending queue and flush it
+ */
+static int flb_output_task_queue_flush_one(struct flb_task_queue *queue)
+{
+    struct flb_task_enqueued *queued_task;
+    int ret;
+
+    ret = mk_list_is_empty(&queue->pending);
+    if (ret == 0) {
+        flb_error("Attempting to pop task queue with no tasks enqueued");
+        return -1;
+    }
+
+    queued_task = mk_list_entry_first(&queue->pending, struct flb_task_enqueued, _head);
+    mk_list_del(&queued_task->_head);
+    mk_list_add(&queued_task->_head, &queue->in_progress);
+    ret = flb_output_task_flush(queued_task->task,
+                                queued_task->out_instance,
+                                queued_task->config);
+
+    return ret;
+}
+
+/*
+ * Will either run or queue running a single task
+ */
+int flb_output_task_singleplex_enqueue(struct flb_task_queue *queue,
+                                       struct flb_task *task,
+                                       struct flb_output_instance *out_ins,
+                                       struct flb_config *config)
+{
+    /* Enqueue task */
+    int ret;
+    ret = flb_output_task_queue_enqueue(queue, task, out_ins, config);
+    if (ret == -1) {
+        return -1;
+    }
+
+    /* Launch task if nothing is running */
+    if (mk_list_is_empty(&out_ins->singleplex_queue->in_progress) == 0) {
+        return flb_output_task_queue_flush_one(out_ins->singleplex_queue);
+    }
+    
+    return 0;
+}
+
+/*
+ * Clear in progress task and flush a single queued task if exists
+ */
+int flb_output_task_singleplex_flush_next(struct flb_task_queue *queue)
+{
+    struct flb_task_enqueued *ended_task;
+
+    /* Remove in progress task */
+    if (!(mk_list_is_empty(&queue->in_progress) == 0)) {
+        ended_task = mk_list_entry_first(&queue->in_progress,
```

Comment:
How do we know that the in progress task is done? Is this based on the assumption that this function is only called once the previous flush finished? If so, I wonder if it would be more clear and readable if there was a separate clean up function?

Reply:
It also seems like this is the only difference between this function and `flb_output_task_queue_flush_one` is that this one also does some clean up. So I think that justifies having one clean up function and one flush next function

Reply:
That sounds good.

Feel free to not read the rest. It's just for me to document how I was thinking
> We know that the inprogress task is done here because the function name is `flb_output_task_singleplex_flush_next` and singleplexing would only flush the next task when the previous task is done.

> I was debating whether or not to structure this part of the code or to have a second function.
I guess the clean up function idea might be a good one.

> The point of having this function separate from flb_output_task_queue_flush_one, was to isolate singleplex logic, and to keep the  flb_output_task_queue_flush_one one generalized. So technically flb_output_task_queue_flush_one could be used to flush multiple tasks from the same queue. I guess that generalization is not really needed.

> This was so that in the future we could implement other schedulers that use the flb_output_task_queue_flush_one and flb_output_task_queue_enqueue without chaging those functions.

Reply:
asdf

Reply:
??

Reply:
oops

Reply:
keep current